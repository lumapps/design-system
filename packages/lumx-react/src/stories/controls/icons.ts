import {
    mdiAbTesting,
    mdiAbjadArabic,
    mdiAccount,
    mdiAccountBox,
    mdiAlert,
    mdiAlertCircle,
    mdiArrowDown,
    mdiArrowUp,
    mdiAtom,
    mdiBee,
    mdiBell,
    mdiBullhornOutline,
    mdiCheck,
    mdiCheckCircle,
    mdiChevronDown,
    mdiChevronLeft,
    mdiChevronRight,
    mdiChevronUp,
    mdiClose,
    mdiCloseCircle,
    mdiDelete,
    mdiDotsHorizontal,
    mdiDragVertical,
    mdiEarth,
    mdiEmail,
    mdiEye,
    mdiFileEdit,
    mdiFlag,
    mdiFolder,
    mdiFolderGoogleDrive,
    mdiFoodApple,
    mdiGoogleCirclesExtended,
    mdiHeart,
    mdiHome,
    mdiImageBroken,
    mdiInformation,
    mdiLink,
    mdiMagnifyMinusOutline,
    mdiMagnifyPlusOutline,
    mdiMenuDown,
    mdiMessageTextOutline,
    mdiMinus,
    mdiOpenInNew,
    mdiPauseCircleOutline,
    mdiPencil,
    mdiPlay,
    mdiPlayCircleOutline,
    mdiPlus,
    mdiRadioboxBlank,
    mdiRadioboxMarked,
    mdiReply,
    mdiSend,
    mdiStar,
    mdiTextBox,
    mdiTextBoxPlus,
    mdiTram,
    mdiTranslate,
    mdiViewList,
} from '@lumx/icons';

// Small selection of mdi icons to
export const iconArgType = {
    control: { type: 'select' },
    options: {
        mdiAbTesting,
        mdiAbjadArabic,
        mdiAccount,
        mdiAccountBox,
        mdiAlert,
        mdiAlertCircle,
        mdiArrowDown,
        mdiArrowUp,
        mdiAtom,
        mdiBee,
        mdiBell,
        mdiBullhornOutline,
        mdiCheck,
        mdiCheckCircle,
        mdiChevronDown,
        mdiChevronLeft,
        mdiChevronRight,
        mdiChevronUp,
        mdiClose,
        mdiCloseCircle,
        mdiDelete,
        mdiDotsHorizontal,
        mdiDragVertical,
        mdiEarth,
        mdiEmail,
        mdiEye,
        mdiFileEdit,
        mdiFlag,
        mdiFolder,
        mdiFolderGoogleDrive,
        mdiFoodApple,
        mdiGoogleCirclesExtended,
        mdiHeart,
        mdiHome,
        mdiImageBroken,
        mdiInformation,
        mdiLink,
        mdiMagnifyMinusOutline,
        mdiMagnifyPlusOutline,
        mdiMenuDown,
        mdiMessageTextOutline,
        mdiMinus,
        mdiOpenInNew,
        mdiPauseCircleOutline,
        mdiPencil,
        mdiPlay,
        mdiPlayCircleOutline,
        mdiPlus,
        mdiRadioboxBlank,
        mdiRadioboxMarked,
        mdiReply,
        mdiSend,
        mdiStar,
        mdiTextBox,
        mdiTextBoxPlus,
        mdiTram,
        mdiTranslate,
        mdiViewList,
    },
};
