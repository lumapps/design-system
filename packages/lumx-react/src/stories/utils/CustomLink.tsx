/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
