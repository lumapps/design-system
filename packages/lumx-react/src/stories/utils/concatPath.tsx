export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
