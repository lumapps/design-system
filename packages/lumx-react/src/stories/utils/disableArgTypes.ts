export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
