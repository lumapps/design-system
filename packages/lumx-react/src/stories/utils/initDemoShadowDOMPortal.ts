/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
