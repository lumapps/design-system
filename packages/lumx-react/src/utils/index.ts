/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
