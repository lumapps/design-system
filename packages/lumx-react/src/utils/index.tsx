export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
