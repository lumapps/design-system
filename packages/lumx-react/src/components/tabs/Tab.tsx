import { Icon, IconProps, Size } from '@lumx/react';
import { CSS_PREFIX } from '@lumx/react/constants';
import { Comp, GenericProps, handleBasicClasses } from '@lumx/react/utils';

import classNames from 'classnames';
import React, { FocusEventHandler, forwardRef, KeyboardEventHandler, ReactNode, useCallback } from 'react';
import { useTabProviderContext } from './state';

/**
 * Defines the props of the component.
 */
export interface TabProps extends GenericProps {
    /** Children are not supported. */
    children?: never;
    /** Icon (SVG path). */
    icon?: IconProps['icon'];
    /** Native id property. */
    id?: string;
    /** Whether the tab is active or not. */
    isActive?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Label content. */
    label: string | ReactNode;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Tab';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = `${CSS_PREFIX}-tabs__link`;

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<TabProps> = {};

/**
 * Tab component.
 *
 * Implements WAI-ARIA `tab` role {@see https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html#rps_label}
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Tab: Comp<TabProps, HTMLButtonElement> = forwardRef((props, ref) => {
    const {
        className,
        disabled,
        icon,
        id,
        isActive: propIsActive,
        isDisabled = disabled,
        label,
        onFocus,
        onKeyPress,
        tabIndex = -1,
        ...forwardedProps
    } = props;
    const state = useTabProviderContext('tab', id);
    const isActive = propIsActive || state?.isActive;

    const changeToCurrentTab = useCallback(() => {
        if (isDisabled) {
            return;
        }
        state?.changeToTab();
    }, [isDisabled, state]);

    const handleFocus: FocusEventHandler = useCallback(
        (event) => {
            onFocus?.(event);
            if (state?.shouldActivateOnFocus) {
                changeToCurrentTab();
            }
        },
        [changeToCurrentTab, onFocus, state?.shouldActivateOnFocus],
    );

    const handleKeyPress: KeyboardEventHandler = useCallback(
        (event) => {
            onKeyPress?.(event);
            if (event.key !== 'Enter') {
                return;
            }
            changeToCurrentTab();
        },
        [changeToCurrentTab, onKeyPress],
    );

    return (
        <button
            ref={ref}
            {...forwardedProps}
            type="button"
            id={state?.tabId}
            className={classNames(className, handleBasicClasses({ prefix: CLASSNAME, isActive, isDisabled }))}
            onClick={changeToCurrentTab}
            onKeyPress={handleKeyPress}
            onFocus={handleFocus}
            role="tab"
            tabIndex={isActive ? 0 : tabIndex}
            aria-disabled={isDisabled}
            aria-selected={isActive}
            aria-controls={state?.tabPanelId}
        >
            {icon && <Icon icon={icon} size={Size.xs} />}
            {label && <span>{label}</span>}
        </button>
    );
});
Tab.displayName = COMPONENT_NAME;
Tab.className = CLASSNAME;
Tab.defaultProps = DEFAULT_PROPS;
