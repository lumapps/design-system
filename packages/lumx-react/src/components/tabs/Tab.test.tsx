import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { build } from 'test-data-bot';

import { mdiCheck } from '@lumx/icons';
import { CommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { CLASSNAME, Tab, TabProps } from './Tab';

/**
 * Define the overriding properties waited by the `setup` function.
 */
type SetupProps = Partial<TabProps>;

/**
 * Defines what the `setup` function will return.
 */
interface Setup extends CommonSetup {
    props: SetupProps;

    /**
     * The <div> element that wraps tab.
     */
    wrapper: Wrapper;
}

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props  The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: SetupProps = {}, shallowRendering: boolean = true): Setup => {
    const props: TabProps = {
        ...propsOverrides,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // noinspection RequiredAttributes
    const wrapper: Wrapper = renderer(<Tab {...props} />);

    return {
        props,
        wrapper,
    };
};

describe(`<${Tab.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { wrapper } = setup();

            expect(wrapper).toMatchSnapshot();
            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use the given props', () => {
            const modifiedPropsBuilder: () => SetupProps = build('props').fields!({
                children: 'Tab Content',
                icon: mdiCheck,
                label: 'Test Tab Label',
            });

            const modifiedProps: SetupProps = modifiedPropsBuilder();

            const { wrapper } = setup({ ...modifiedProps });
            expect(wrapper).toMatchSnapshot();
        });

        it('should use the given props to add classes', () => {
            const modifiedPropsBuilder: () => SetupProps = build('props').fields!({
                isActive: true,
                isDisabled: true,
            });

            const modifiedProps: SetupProps = modifiedPropsBuilder();

            const { wrapper } = setup({ ...modifiedProps });

            Object.keys(modifiedProps).forEach((prop: string) => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                );
            });
        });
    });

    // 3. Test events.
    describe('Events', () => {
        const onTabClick: jest.Mock = jest.fn();

        beforeEach(() => {
            onTabClick.mockClear();
        });

        it('should trigger `onTabClick` when clicked', () => {
            const { wrapper } = setup({ index: 7, onTabClick }, false);

            wrapper.simulate('click');
            expect(onTabClick).toHaveBeenCalledWith({ event: jasmine.any(Object), index: 7 });
        });

        it('should trigger `onTabClick` when pressing `enter` key', () => {
            const { wrapper } = setup({ index: 9, onTabClick }, false);

            wrapper.simulate('keypress', { keyCode: 13 });
            expect(onTabClick).toHaveBeenCalledWith({ event: jasmine.any(Object), index: 9 });
        });

        it('should not trigger `onTabClick` when pressing any other key', () => {
            const { wrapper } = setup({ index: 10, onTabClick }, false);

            wrapper.simulate('keypress', { keyCode: 12 });
            expect(onTabClick).not.toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, {}, { className: CLASSNAME });
});
