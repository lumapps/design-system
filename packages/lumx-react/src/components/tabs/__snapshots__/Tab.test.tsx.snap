// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab> Snapshots and structure should render 1`] = `
<button
  aria-controls="tab-panel-1"
  aria-selected={false}
  className="lumx-tabs__link"
  id="tab-1"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={-1}
  type="button"
>
  <span>
    Test Tab Label
  </span>
</button>
`;

exports[`<Tab> Snapshots and structure should render active state 1`] = `
<button
  aria-controls="tab-panel-3"
  aria-selected={true}
  className="lumx-tabs__link lumx-tabs__link--is-active"
  id="tab-3"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={0}
  type="button"
>
  <span>
    Test Tab Label
  </span>
</button>
`;

exports[`<Tab> Snapshots and structure should render icon 1`] = `
<button
  aria-controls="tab-panel-2"
  aria-selected={false}
  className="lumx-tabs__link"
  id="tab-2"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={-1}
  type="button"
>
  <Icon
    icon="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
    size="xs"
  />
  <span>
    Test Tab Label
  </span>
</button>
`;
